### PR TITLE
Implement multipliers for fault transmissibility specification

### LINKED ISSUE #1124: [Feature] Fault permeability specification
**What is the requested feature?**
Faults can now be identified as node sets.  We would like to enable the ability to treat faults as either flow conduits or flow barriers.  There are two cases to consider:
* Flow barrier: we just need to cut the transmissibility graph across faces connected through a fault face
* Flow conduit: we need to add a FaceElement to the fault face, and be able to specify across fault and along fault perms

We had some good discussion on this at the last dev mtg. @rrsettgast and @klevzoff, could you summarize your design suggestions?

### LINKED ISSUE #1191: [Bug] integratedTests/compositionalMultiphaseFlow/CO2_flux_3d.xml
There is a typo when defining source and sink
Should be 26.99, not 27.99
Should be  3.01 , not 2.01

### PR DESCRIPTION
**Pull request by francoishamon**
This PR implements multipliers to specify transmissibility for cross-fault flow. The multipliers are specified using a `FieldSpecification` block in the xml file. 
- For `SinglePhaseFVM`, the stencil weights are modified in `TwoPointFluxApproximation.cpp`.
- For `SinglePhaseHybridFVM`, the transmissibility matrix is modified on the fly in `HybridFVMInnerProduct.hpp`. 

I modified two existing integrated tests and added one for hybrid FVM (which also tests other things not tested for this scheme before, like multiple element regions with different properties).

Resolves #1124 for cross-fault flow. 
It seems to work for the specification of impermeable faults in the HI24L case that motivated the issue.
Thanks @klevzoff for the advice.

Fixes #1191 

Related to https://github.com/GEOSX/integratedTests/pull/102.

### CODE REVIEW

**Review by klevzoff — APPROVED** [review]

File: src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp [hunk 1]
```diff
@@ -95,7 +98,7 @@ void TwoPointFluxApproximation::computeCellStencil( MeshLevel & mesh ) const
   forAll< serialPolicy >( faceManager.size(), [=, &stencil]( localIndex const kf )
   {
     // Filter out boundary faces
-    if( elemList[kf][0] < 0 || elemList[kf][1] < 0 )
+    if( elemList[kf][0] < 0 || elemList[kf][1] < 0 || transMultiplier[kf] < 1e-16 )
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp]: Done

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseHybridFVMKernels.hpp [hunk 2]
```diff
@@ -65,59 +65,59 @@ struct AssemblerKernelHelper
...
+   * @param[in] esr index of this element's subregion
+   * @param[in] ei index of this element
    * @param[in] elemRegionList face-to-elemRegions map
    * @param[in] elemSubRegionList face-to-elemSubRegions map
    * @param[in] elemList face-to-elemIds map
    * @param[in] regionFilter set containing the indices of the target regions
    * @param[in] elemToFaces elem-to-faces maps
-   * @param[in] mob the mobilities in the domain (non-local)
-   * @param[in] dMob_dp the derivatives of the mobilities in the domain wrt cell-centered pressure (non-local)
+   * @param[in] mobilities the mobilities in the domain (non-local)
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseHybridFVMKernels.hpp]: Thanks for catching this, I changed the Doxygen documentation here.

File: src/coreComponents/finiteVolume/FluxApproximationBase.cpp [hunk 3]
```diff
@@ -90,6 +90,14 @@ void FluxApproximationBase::RegisterDataOnMesh( Group * const meshBodies )
       {
         registerBoundaryStencil( stencilGroup, setName );
       } );
+
+      FaceManager & faceManager = *mesh.getFaceManager();
+      faceManager.registerWrapper< array1d< real64 > >( viewKeyStruct::transMultiplierString )->
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/finiteVolume/FluxApproximationBase.cpp]: Sorry I was confused, I realize now that I had misunderstood what I had to do. Now I register the field in `FluxApproximationBase::RegisterDataOnMesh` as `m_coeffName + viewKeyStruct::transMultiplierString`, and I removed the `fsManager.Apply()` call here. I also added the `initialCondition="1"` in the integrated tests (and everything still works). Thanks

File: src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp [hunk 4]
```diff
@@ -302,6 +306,25 @@ struct QTPFACellInnerProductKernel
...
+
+    // placeholder before we can implement the hybrid treatment of trans multipliers
+    for( localIndex i = 0; i < NF; ++i )
+    {
+      real64 const mult = transMultiplier[elemToFaces[i]];
+      for( localIndex j = 0; j < NF; ++j )
+      {
+        transMatrix[i][j] *= mult;
+      }
+    }
```
Comment (francoishamon) [inline src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: @corbett5 Here what I would like to ultimately do is to implement correctly the transmissibility multipliers for the hybrid FVM scheme. For this, I would need to compute the inverse of `transMatrix` in the kernel to do something like:
```
transMatrix = inv( inv( transMatrix ) + D )
```
where `transMatrix` is the local transmissibility matrix of size `NF x NF` and `D` is a diagonal matrix of size `NF x NF` involving the multipliers. For this, I would need `LvArray::tensorOps::invert< NF >` for `NF = 4`, `NF = 5`, and `NF = 6` (and maybe for more faces later). Based on what I saw in `LvArray/src/fixedSizeSquareMatrixOpsImpl.hpp` I think this is not supported yet but if you ever implement it I think this would be a possible use case (it is very low priority).

Comment (corbett5) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: Cool! As much as I would like to I don't have time to work on this, could you make an issue in LvArray so we can keep track of the request? Adding a general matrix inversion would be a departure from the current supported methods which are all closed form but that doesn't mean it's not a good idea. If you know of an algorithm that might work nicely for small sizes add it to the issue. Or if you think `(A^-1 + D)^-1` is a common pattern there might be a better way to specifically calculate that.

Comment (francoishamon) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: Thinking about it more, I think I can use a workaround to compute `(A^-1 + D)^-1` without taking any inverse. Here, `A^-1` is a symmetric definite positive matrix, and `D` is a diagonal matrix. I can write this as `(A^-1 + D_1 + ... + D_{NF})^-1` where `D_i` are rank-one matrices. Then I can apply the Sherman-Morrison formula multiple times (for example, four times if `NF = 4`) to arrive at the result with matrix products and matrix sums only, without taking any inverse (I need to try to convince myself). 

I am going to give it a try and I'll let you know. This may be very specific to my problem though. If I don't arrive at anything satisfactory I will post an issue in LvArray. Thanks!

Comment (corbett5) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: Nice! I did a little toying around it seems like it should work. It would still have complexity of `n^3` which is the same as standard inversion but it is *probably* faster.

Comment (rrsettgast) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: @francoishamon There are a lot of zeros in there since the perturbation vectors are equal. And of course T is symmetric. It will be certainly faster than generic inversion. I think you perhaps make further simplifications aside from the obvious `MatrixMatrixMatrix` and `VectorMatrixVector products`. I can clean up the Mathematica sheet and send it with you if you would like to pursue it.

Comment (corbett5) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: The `VectorMatrixVector` product in the denominator works out to just a single multiplication for one iteration. The `MatrixMatrixMatrix` product in the numerator works out to O(N^2) multiplications.

Comment (francoishamon) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: @rrsettgast Yes definitely that would be great if you could send it to me and I will implement it in this PR.

Thank you @rrsettgast @corbett5 for the help.

Comment (rrsettgast) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: @francoishamon I think I was chasing a ghost. I think the direct application of Sherman-Morrison taking advantage of the fact that `T(i)` is symmetric, `Outer( D(i), D(i) )` has a single nonzero value, and `D(i) T(i-1) D(i)` is a single FMA operation is the best you will do.

Comment (rrsettgast) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp:327]: I think it would look something like:
```suggestion
    // placeholder before we can implement the hybrid treatment of trans multipliers
    for( localIndex k = 0; k < NF; ++k )
    {
      real64 const mult = transMultiplier[elemToFaces[i]];
      real64 Td[NF];
      for( localIndex i = 0; i < NF; ++i )
      {
        Td[i] = T[k][i] * mult;
      }
      real64 const invDenom = 1.0 / ( 1.0 + Td[k] * mult );
      for( localIndex i = 0; i < NF; ++i )
      {
         for( localIndex j = 0; j < NF; ++j )
        {
          transMatrix[i][j] += Td[i]*Td[j]*invDenom ;
        }
     }
   }
```

Comment (francoishamon) [reply src/coreComponents/finiteVolume/HybridFVMInnerProduct.hpp]: @rrsettgast Thank you for the suggestion it works great. To compute the `(A^-1 + D)^-1` (with a diagonal `D` involving the TPFA transmissibility and the multiplier), the formula does exactly what I need.

**Comment by francoishamon** [comment]
This PR passes tests on Quartz and Lassen on integratedTests PR https://github.com/GEOSX/integratedTests/pull/102 and is ready to be merged.

**Review by rrsettgast — APPROVED** [review]
